presto-main/src/test/java/io/prestosql/execution/resourcegroups/TestResourceGroups.java

```diff
@@ -321,14 +323,341 @@ public void testHardCpuLimit()
+        q1.setMemoryUsage(new DataSize(1, BYTE));
+
+        // A new incoming query q3 will get queued since cached usage is still 2B.
+        MockResourceUpdatingQueryExecution q3 = new MockResourceUpdatingQueryExecution();
+        child.run(q3);
+        assertEquals(q3.getState(), QUEUED);
+
+        // q2 and q3 start running when cached usage is updated and queued queries are processed.
+        root.updateGroupsAndProcessQueuedQueries();
+        assertEquals(q2.getState(), RUNNING);
+        assertEquals(q3.getState(), RUNNING);
+    }
+
+    @Test(timeOut = 10_000)
+    public void testMemoryUsageUpdateAtQueryCompletion()
+    {
+        RootInternalResourceGroup root = new RootInternalResourceGroup("root", (group, export) -> {}, directExecutor());
+        InternalResourceGroup child = root.getOrCreateSubGroup("child");
+
+        // Set the common configurations for the two groups
+        List<InternalResourceGroup> groups = ImmutableList.of(root, child);
+        groups.forEach(group -> {
+            group.setHardConcurrencyLimit(100);
+            group.setMaxQueuedQueries(100);
+            group.setSoftMemoryLimit(new DataSize(1.5, BYTE));
+        });
+
+        MockResourceUpdatingQueryExecution q1 = new MockResourceUpdatingQueryExecution();
+        child.run(q1);
+        assertEquals(q1.getState(), RUNNING);
+
+        // Invoke updateGroupsAndProcessQueuedQueries to make sure that 2B memory usage is cached
+        q1.setMemoryUsage(new DataSize(2, BYTE));
+        root.updateGroupsAndProcessQueuedQueries();
+
+        // Completion of q1 should reduce the cached memory usage to 0B.
+        q1.complete();
+
+        // q2 starts running since the current memory usage 0B is less than the limit 1.5B.
+        MockResourceUpdatingQueryExecution q2 = new MockResourceUpdatingQueryExecution();
+        child.run(q2);
+        assertEquals(q2.getState(), RUNNING);
+    }
+
+    /**
+     * A test for correct cpu usage update aggregation and propagation in non-leaf nodes. It uses in a multi
+     * level resource group tree, with non-leaf resource groups having more than one child.
+     */
+    @Test(timeOut = 10_000)
+    public void testRecursiveCpuUsageUpdate()
+    {
+        RootInternalResourceGroup root = new RootInternalResourceGroup("root", (group, export) -> {}, directExecutor());
+        InternalResourceGroup rootChild1 = root.getOrCreateSubGroup("rootChild1");
+        InternalResourceGroup rootChild2 = root.getOrCreateSubGroup("rootChild2");
+        InternalResourceGroup rootChild1Child1 = rootChild1.getOrCreateSubGroup("rootChild1Child1");
+        InternalResourceGroup rootChild1Child2 = rootChild1.getOrCreateSubGroup("rootChild1Child2");
+
+        // Set the same values in all the groups for some configurations
+        List<InternalResourceGroup> groups = ImmutableList.of(root, rootChild1, rootChild2, rootChild1Child1, rootChild1Child2);
+        groups.forEach(group -> {
```

Comment:
you can use `Stream.of(root, rootChild1, rootChild2, rootChild1Child1, rootChild1Child2).forEach(groups -> {`